### ISSUE TITLE
Using Babel's `format_number` as a custom filter give strange results!

### ISSUE THREAD

**Issue by stucox** [grow/grow#302]
We have:

``` python
# extensions/filters.py
from babel.numbers import format_number

class Filters(Extension):

    def __init__(self, environment):
        super(Filters, self).__init__(environment)
        environment.filters['format_number'] = format_number
```

``` yaml
# podspec.yaml
...
extensions:
  jinja2:
    - extensions.filters.Filters
```

``` html
<!-- views/template.html -->
{{ 1000|format_number(locale='en-US') }}
```

With Grow 0.0.66 globally installed (via `curl https://install.growsdk.org | bash`), for some reason when the template renders, the number is multiplied by 10?!

```
10,000
```

I tried to debug by checking out tag `0.0.66` of the Grow repo, running `make develop`, going back to my project, creating a virtualenv and installing Grow into it with `pip install ../grow` (the Grow checkout is alongside my project)… and the number is rendered correctly.

This happens consistently for other numbers, other locales and on other people’s machines: globally installed Grow multiplies by 10; `pip install`’d Grow doesn't.

😱 

I’m out of ideas of how to investigate this one… is there any way to run the installed Grow with `pdb`?

**Comment by stucox** [comment]
Ok, it appears it goes wrong @ [babel/numbers.py#L571](https://github.com/python-babel/babel/blob/2.2.0/babel/numbers.py#L571):

```python
 value = value.scaleb(self.scale)
```

With the installed Grow 0.0.66, `self.scale` is `1`, and hence `value` is raised to `E+1` (i.e. multiplies by 10). With a dev install of Grow, `self.scale` is `0`, and hence `value` doesn’t change.

It appears this difference is because the CLDR data bundled with Grow 0.0.66 doesn’t match the version bundled with Babel 2.2.0 (which comes from http://unicode.org/Public/cldr/28/core.zip)

I’m not familiar with how Grow releases are built, so not sure how this would happen… any ideas @jeremydw?

**Comment by jeremydw** [comment]
That's weird -- thanks for the report. Well, it's bizarre that `self.scale` is 1 because it looks like the valid values for it are either 2, 3, or 0 (https://github.com/python-babel/babel/blob/2.2.0/babel/numbers.py#L557). Grow's release is built using `make release` -- which is a convoluted process that freezes Grows source code into an executable for distribution. CLDR data is distributed as data files, so its possible that the frozen code cannot read the data files properly (wonder if this problem is also related to https://github.com/grow/grow/issues/291).

**Comment by stucox** [comment]
In 2.2.0 `self.scale` is 0, 2 or 3 and represents an exponent, but in 2.1.1 `self.scale` was [1, 100 or 1000](https://github.com/python-babel/babel/blob/2.1.1/babel/numbers.py#L594) and represented a multiple.

I’m not sure if the data Babel packages is adapted from the original CLDR data, but my suspicion is that binary data packaged for 2.1.1 (`self.scale` = 1) was bundled with the 2.2.0 source code (treated as an exponent rather than multiple) — this would certainly explain it.

**Comment by jeremydw** [comment]
After some painful debugging I've found that the fix for this is part of PyInstaller v3.0. See here -- https://github.com/pyinstaller/pyinstaller/releases/tag/3.0 ("hook fixes for babel").

The build machine has used PyInstaller 2.1 because versions >2.1 have had problems. Even the most recent stable version has had build problems (https://github.com/pyinstaller/pyinstaller/issues/1586 -- users recommend using 2.1 instead).

I was able to both fix the frozen Babel issue and the PyInstaller issue by pinning PyInstaller to a specific version on the build machine (`pip install git+https://github.com/pyinstaller/pyinstaller.git#b78bfe530cdc2904f65ce098bdf2de08c9037abb`) and rebuilding.

I'll remedy this with a few things by updating the `Makefile` with a new rule for installing this specific version of PyInstaller on the build machine.

I also have one more question related to this -- for this particular function, and also for https://github.com/grow/grow/issues/291 it seems like it might be a good idea to make these functions available as default template filters. Thoughts? We can do that in a future version.

**Comment by Zoramite** [comment]
@stucox In [version `0.1.0`](https://github.com/grow/grow/releases/tag/0.1.0) we have added in more of the babel number/currency functionality as built-in jinja filters so this should now work natively: `{{12345|number}}` and defaults to the current document's locale unless explicitly overridden.
